python-checks/src/main/java/org/sonar/python/checks/NoReRaiseInExitCheck.java

```diff
@@ -0,0 +1,137 @@
+
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.python.api.tree.ExceptClause;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.HasSymbol;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Parameter;
+import org.sonar.plugins.python.api.tree.ParameterList;
+import org.sonar.plugins.python.api.tree.RaiseStatement;
+import org.sonar.plugins.python.api.tree.SubscriptionExpression;
+import org.sonar.plugins.python.api.tree.Tree;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+@Rule(key="S5706")
+public class NoReRaiseInExitCheck extends PythonSubscriptionCheck {
+
+  private static class RaiseVisitor extends BaseTreeVisitor {
+    private Symbol caughtException;
+    private Symbol packedParameter;
+
+    private List<RaiseStatement> nonCompliantRaises = new ArrayList<>();
+
+    public RaiseVisitor(Symbol caughtException, Symbol packedParameter) {
+      this.caughtException = caughtException;
+      this.packedParameter = packedParameter;
+    }
+
+    @Override
+    public void visitExceptClause(ExceptClause exceptClause) {
+      // Intentionally empty - we do not want to enter except blocks.
+    }
+
+    @Override
+    public void visitRaiseStatement(RaiseStatement pyRaiseStatementTree) {
+      if (pyRaiseStatementTree.expressions().isEmpty()) {
+        nonCompliantRaises.add(pyRaiseStatementTree);
+        return;
+      }
+
+      Expression raisedException = pyRaiseStatementTree.expressions().get(0);
+      if (raisedException instanceof HasSymbol && Objects.equals(((HasSymbol) raisedException).symbol(), caughtException)) {
+        nonCompliantRaises.add(pyRaiseStatementTree);
+      }
+
+      if (raisedException.is(Tree.Kind.SUBSCRIPTION)) {
+        SubscriptionExpression subscriptionExpression = (SubscriptionExpression) raisedException;
+        Expression objectExpression = subscriptionExpression.object();
+
+        if (objectExpression instanceof HasSymbol && Objects.equals(((HasSymbol) objectExpression).symbol(), packedParameter)) {
+          nonCompliantRaises.add(pyRaiseStatementTree);
+        }
+      }
+    }
+  }
+
+  private static Symbol extractPackedParameter(ParameterList parameterList) {
+    if (parameterList.nonTuple().size() != 2) {
+      return null;
+    }
+
+    Parameter parameter = parameterList.nonTuple().get(1);
+    if (parameter.starToken() == null) {
+      return null;
+    }
+
+    Name name = parameter.name();
+    if (name != null) {
+      return name.symbol();
+    }
+
+    return null;
+  }
+
+  private static Symbol extractCaughtExceptionParameter(ParameterList parameterList) {
+    if (parameterList.nonTuple().size() != 4) {
```

Comment:
I believe I can create a case with some Python 2 tuple parameters which make that test not quite correct. However, I think it's not worth trying to cover such corner cases.